rfcs/0014-opendal.md

```diff
@@ -0,0 +1,196 @@
+
+One notable implementation detail is the special handling of `list()` operations. Because the list API returns a stream and cannot be easily retried mid-iteration, the current implementation collects the entire stream into a Vec before attempting the operation. This ensures atomicity—either the entire list succeeds or the entire operation is retried.
+
+OpenDAL provides a [`RetryLayer`](https://opendal.apache.org/docs/rust/opendal/layers/struct.RetryLayer.html) with configurable retry policies that can potentially replace our custom implementation. It retries on errors which `Error::is_temporary` returns true.
+
+However, we may want to customize the retry behavior to match our current implementation. For example, when encountering an `AlreadyExists` error, we may want to read the object metadata (via a `stat()` call) to determine whether to return success (if we created the object and it already exists, making the operation idempotent) or return a conflict error (if someone else created it).
+
+Since OpenDAL's built-in `RetryLayer` did not expose an interface to allow us customize the retry behavior, we could consider implementing a customized `RetryLayer` in our repository. This approach would allow us to preserve our specialized retry logic (such as the idempotent `AlreadyExists` handling) while still benefiting from OpenDAL's ecosystem and avoiding the need to implement the entire `ObjectStore` trait as we do currently.
+
+<!-- TOC --><a name="cachedobjectstore"></a>
+### CachedObjectStore
+
+The `CachedObjectStore` (`cached_object_store/`) is the most complex extension of the `ObjectStore` trait in the current codebase. It provides transparent chunked caching with sophisticated cache management, including:
+
+1. **Chunked caching**: Objects are split into align sized parts and cached independently. This allows partial reads to be served from cache without fetching entire objects.
+2. **Cache admission control**: An `AdmissionPicker` determines which objects should be cached based on access patterns, preventing cache pollution from one-time reads.
+3. **Pick-of-2 eviction strategy**: When cache size exceeds the limit, the evictor uses a pick-of-2 algorithm to approximate LRU—it randomly selects two cached files, compares their last access times, and evicts the older one. This requires a background task to periodically track the files list in the memory.
+
+At the time of writing, OpenDAL does not provide a built-in cache layer that matches our requirements. There was an issue that aims to add a cache layer for opendal in <https://github.com/apache/opendal/issues/5678>, however, this issue does not have much progress. There're still exists some community maintained cache layers, such as <https://github.com/r3v2d0g/opendal-foyer>.
+
+In the longer term, we hope the complexities around caching can be maintained with community force. Although the cache layer still haven't been built in from OpenDAL, we can push the force together to build it.
```

Comment:
I'll be honest: https://github.com/apache/opendal/issues/5678 doesn't give me a lot of hope. It feels the same as https://github.com/apache/arrow-rs-object-store/issues/14. One could reasonably ask why we should expect caching to get done faster in OpenDAL than in the object_store GH issue. In both cases there seems to be a high-level awareness that this feature would be nice, but no real engagement even if users are pushing for it (and submitting RFCs to opendal or code to object_store).